bdk_electrum/src/lib.rs

```diff
@@ -0,0 +1,583 @@
+                // Just recursively call (this should never happen).
+                return prepare_update(client, local_chain);
+            }
+        }
+    }
+
+    Ok(update)
+}
+
+fn populate_with_outpoints(
+    client: &Client,
+    update: &mut SparseChain,
+    outpoints: &mut impl Iterator<Item = OutPoint>,
+) -> Result<HashMap<Txid, Transaction>, InternalError> {
+    let tip = update
+        .latest_checkpoint()
+        .expect("update must atleast have one checkpoint");
+
+    let mut full_txs = HashMap::new();
+    for outpoint in outpoints {
+        let txid = outpoint.txid;
+        let tx = client.transaction_get(&txid)?;
+        debug_assert_eq!(tx.txid(), txid);
+        let txout = match tx.output.get(outpoint.vout as usize) {
+            Some(txout) => txout,
+            None => continue,
+        };
+
+        // attempt to find the following transactions (alongside their chain positions), and
+        // add to our sparsechain `update`:
+        let mut has_residing = false; // tx in which the outpoint resides
+        let mut has_spending = false; // tx that spends the outpoint
+        for res in client.script_get_history(&txout.script_pubkey)? {
+            if has_residing && has_spending {
+                break;
+            }
+
+            // skip if we have already added the tx to our `update`
+            if full_txs.contains_key(&res.tx_hash) {
+                if res.tx_hash == txid {
+                    has_residing = true;
+                } else {
+                    has_spending = true;
+                }
+                continue;
+            }
+
+            let res_tx = if res.tx_hash == txid {
+                has_residing = true;
+                tx.clone()
+            } else {
+                let res_tx = client.transaction_get(&res.tx_hash)?;
+                if !res_tx
+                    .input
+                    .iter()
+                    .any(|txin| txin.previous_output == outpoint)
+                {
+                    continue;
+                }
+                has_spending = true;
+                res_tx
+            };
+            full_txs.insert(res.tx_hash, res_tx);
+
+            let tx_height = match res.height {
+                // Electrum server API specifies that height <= 0 is considered unconfirmed
+                h if h <= 0 => {
+                    debug_assert!(
+                        h == 0 || h == -1,
+                        "unexpected height ({}) from electrum server",
+                        h
+                    );
+                    TxHeight::Unconfirmed
+                }
+                h => {
+                    let h = h as u32;
+                    if h > tip.height {
+                        TxHeight::Unconfirmed
+                    } else {
+                        TxHeight::Confirmed(h)
+                    }
+                }
+            };
+
+            if let Err(failure) = update.insert_tx(txid, tx_height) {
```

Comment:
Can you double check this line? It seems to me that `txid` refers to `tx`, but `tx_height` might not, as it might refer to the tx spending

Reply:
`update` is a `SparseChain`, so it's `txid` here!

Reply:
I'm not sure I understand :(

What I meant is: `txid` is referring to the initial `tx`, but `tx_height` refers to `res`, and `res` is one of the txs returned by `client.script_get_history(&txout.script_pubkey)?`. So I'm wondering if you're inserting a `txid` with a confirmation height that belongs to a different transaction

Reply:
`res_tx` should always be the same tx as that of `tx` at that point (so I hope).

Check lines from `412`:
```rust
            let res_tx = if res.tx_hash == txid {
                has_residing = true;
                tx.clone()
            } else {
                let res_tx = client.transaction_get(&res.tx_hash)?;
                if !res_tx
                    .input
                    .iter()
                    .any(|txin| txin.previous_output == outpoint)
                {
                    continue;
                }
                has_spending = true;
                res_tx
            };
```

Reply:
The else branch is returning the transaction spending from `tx` I think
```
            } else {
                let res_tx = client.transaction_get(&res.tx_hash)?;
                if !res_tx
                    .input
                    .iter()
                    .any(|txin| txin.previous_output == outpoint)
                {
                    continue;
                }
                has_spending = true;
                res_tx
            };
```

Reply:
Oh damn, good point I completely missed that thanks!

Reply:
Fixed in 97f5bec7266138567c49e6f8f343e86fe0227464